objgraph.py

```diff
@@ -355,6 +356,112 @@ def show_growth(limit=10, peak_stats=None, shortnames=True, file=None,
+
+    limit: the maximum number of rows that you want to print data for
+
+    _state: a dictionary which stores old, current, and new_ids in memory.
+    Never pass in this argument, it is used by the function to store ids in
+    memory.
+
+    if ``skip_update`` is True, the dictionary new_ids
+    will be returned from when the function was last run without examining the
+    objects currently in memory.
+
+    The caveats documented in :func:`growth` apply.
+
+    When one gets new_ids from get_new_ids, one can use at_addrs to
+    get a list of those objects. Then one can iterate over the new
+    objects, print out what they are, and call show_chain to see where they
+    are referenced.
+
+    Example:
+
+        >>> _ = get_new_ids() # store current objects in _state
+        >>> _ = get_new_ids() # curent_ids become old_ids in _state
+        >>> a = [0, 1, 2] # list we don't know about
+        >>> b = [3, 4, 5] # list we don't know about
+        >>> new_ids = get_new_ids(limit=3) # we see new lists
+        ======================================================================
+        Type                    Old_ids  Current_ids      New_ids Count_Deltas
+        ======================================================================
+        list                        324          326           +3           +2
+        dict                       1125         1125           +0           +0
+        wrapper_descriptor         1001         1001           +0           +0
+        ======================================================================
+        >>> new_lists = at_addrs(new_ids['list'])
+        >>> a in new_lists
+        True
+        >>> b in new_lists
+        True
+    """
+    if not _state:
+        _state['old'] = collections.defaultdict(set)
+        _state['current'] = collections.defaultdict(set)
+        _state['new'] = collections.defaultdict(set)
+    new_ids = _state['new']
+    if skip_update:
+        return new_ids
+    old_ids = _state['old']
+    current_ids = _state['current']
+    gc.collect()
+    objects = gc.get_objects()
+    for class_name in old_ids:
+        old_ids[class_name].clear()
+    for class_name, ids_set in current_ids.items():
+        old_ids[class_name].update(ids_set)
+    for class_name in current_ids:
+        current_ids[class_name].clear()
+    for o in objects:
+        class_name = type(o).__name__
+        id_number = id(o)
+        current_ids[class_name].add(id_number)
+    for class_name in new_ids:
+        new_ids[class_name].clear()
+    rows = []
+    for class_name in current_ids:
+        num_old = len(old_ids[class_name])
+        num_current = len(current_ids[class_name])
+        if num_old == 0 and num_current == 0:
+            # remove the key from our dicts if we don't have any old or
+            # curent class_name objects
+            del old_ids[class_name]
+            del current_ids[class_name]
```

Comment:
You're modifying a dictionary while iterating over it.  This can cause problems (RuntimeError: dictionary changed size during iteration).